api/lib/flight_job_script_api/system_command.rb

```diff
@@ -187,75 +182,151 @@ def run(&block)
               out: out_write, err: err_write, in: in_read
             }
             # NOTE: Keep the log before the exec for timing purposes
-            FlightJobScriptAPI.logger.info("Executing (#{user}): #{cmd.join(' ')}")
+            FlightJobScriptAPI.logger.debug("Executing (#{user}): #{stringified_cmd}")
             Kernel.exec(env, *cmd, **opts)
           end
         end
 
-        # Close the child pipes
-        out_write.close
-        err_write.close
-        in_read.close
+        wait_for_command
+        log_command
+      end
+    end
+
+    private
+
+    def passwd
+      @passwd ||= Etc.getpwnam(user)
+    end
 
-        # Wait for the command to finish within the timelimit
-        signal = 'TERM'
+    def with_pipes_and_mutex
+      mutex.synchronize do
         begin
-          FlightJobScriptAPI.logger.debug("Waiting for pid: #{pid}")
-          _, status = Timeout.timeout(FlightJobScriptAPI.app.config.command_timeout) do
-            Process.wait2(pid)
-          end
-        rescue Timeout::Error
-          FlightJobScriptAPI.logger.error("Sending #{signal} to: #{pid}")
-          Process.kill(-Signal.list[signal], pid)
-          signal = 'KILL'
-          retry
+          # Create the pipes
+          @out_read,  @out_write = IO.pipe
+          @err_read,  @err_write = IO.pipe
+          @in_read,   @in_write  = IO.pipe
+
+          # Yield control
+          yield if block_given?
+
+        ensure
+          # Close the pipes
+          out_read.close  if out_read   && !out_read.closed?
+          out_write.close if out_write  && !out_write.closed?
+          err_read.close  if err_read   && !err_read.closed?
+          err_write.close if err_write  && !err_write.closed?
+          in_read.close   if in_read    && !in_read.closed?
+          in_write.close  if in_write   && !in_write.closed?
         end
+      end
+    end
 
-        # Store the results
-        if status.exitstatus
-          self.exitstatus = status.exitstatus
-
-        # The exit status from signals aren't always set, instead they can be inferred
-        elsif status.signaled?
-          FlightJobScriptAPI.logger.warn "Inferring exit code from signal #{Signal.signame(status.termsig)} (pid: #{pid})"
-          self.exitstatus = status.termsig + 128
+    def wait_for_command
+      # Close the child pipes
+      out_write.close
+      err_write.close
+      in_read.close
+
+      # Various control variables
+      signal = 'TERM'
+      loop_stdout = true
+      loop_stderr = true
+      start = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      step = FlightJobScriptAPI.config.command_timeout / 1000.to_f
```

Comment:
I'm a bit concerned about running the entire code block within a `Timeout`. This opens the possibility that an error is raised during one of the `read_nonblocks` preventing that bit of input being lost.

---

At best, the process will receive `SIGTERM` before going back and reading the rest of `stdout`/`stderr`. The child would exit with 143 and generate some funky logs.

At worst the child has already exited 0, and the IO buffer just needs to be cleared. This could cause a "successful" command with a broken `stdout` and lead to a JSON parsing error.

---

In general I wanted to shy away from raising errors during `read_nonblock`. The same would apply to using `read_partial` with a `Timeout`.

Reply:
> I'm a bit concerned about running the entire code block within a Timeout.

This sounds like you're trying to interleave waiting for the process and reading its output.  Which of course is exactly what is needed.  Separate threads for waiting and reading would be a better way to interleave them.

> This opens the possibility that an error is raised during one of the `read_nonblocks`; causing a bit of output to be lost.

I'm not sure what exceptions you're concerned with, but `read_partial` only raises `EOFError`.

---

Can you explain the following in more detail.  Right now it sounds like a timing bug waiting to happenn.

> At best, the process will receive SIGTERM before going back and reading the rest of stdout/stderr. The child would exit with 143 and generate some funky logs.
>
> At worst the child has already exited 0, and the IO buffer just needs to be cleared. This could cause a "successful" command with a broken stdout and lead to a JSON parsing error.

Reply:
Suppose something like this:

```
Timeout.timeout(period) do
end
```